### ISSUE TITLE
Add support for CollectionAttribute<T> and CollectionDefinitionAttribute<T>

### ISSUE THREAD

**Issue by jeffward01** [xunit/xunit#2629]
I reviewed the code and I do not think this feature exists, but I just wanted to ask anyways.

# Question

Is it possible to use `[CollectionDefinition("myCollectionFixtureName")]` with generic attributes?

[Here is the current documentation ](https://xunit.net/docs/shared-context)for `ICollectionFixture<>`

### Current 

```csharp

[CollectionDefinition("Database collection")]
public class DatabaseCollection : ICollectionFixture<DatabaseFixture>
{
    // This class has no code, and is never created. Its purpose is simply
    // to be the place to apply [CollectionDefinition] and all the
    // ICollectionFixture<> interfaces.
}



[Collection("Database collection")]
public class DatabaseTestClass1
{
    DatabaseFixture fixture;

    public DatabaseTestClass1(DatabaseFixture fixture)
    {
        this.fixture = fixture;
    }
}

````

### Expected C# 11 functionality

```csharp

[CollectionDefinition<DatabaseCollection>]
public class DatabaseCollection : ICollectionFixture<DatabaseFixture>
{
    // This class has no code, and is never created. Its purpose is simply
    // to be the place to apply [CollectionDefinition] and all the
    // ICollectionFixture<> interfaces.
}



[Collection<DatabaseCollection>]
public class DatabaseTestClass1
{
    DatabaseFixture fixture;

    public DatabaseTestClass1(DatabaseFixture fixture)
    {
        this.fixture = fixture;
    }
}

```

-----

## Summary
* Is this possible?
*  Is this something that I can open an issue for and myself _(alas... if i have time)_ or someone else can  implement?

**Comment by jeffward01** [comment]
I also opened a discussion here that has not gotten much traction: https://github.com/xunit/xunit/discussions/2627

**Comment by benmccallum** [comment]
Just found myself wondering this too, if I ever get time I could explore this.

**Comment by benmccallum** [comment]
I think it'd be good to rename this issue @jeffward01 to something easier to find like "Use C#11's generic attributes for xunit attributes" as this would be something that'd be useful for a quite a few attributes I imagine, e.g. `[ClassData<T>]`

**Comment by bradwilson** [comment]
As an FYI, I did a little exploration around this to see if it could be done in v2, and it cannot, because we don't have the ability to get generic types from [`IAttributeInfo`](https://github.com/xunit/abstractions.xunit/blob/main/src/xunit.abstractions/Reflection/IAttributeInfo.cs).

This can be supported in v3.

**Comment by JamesTerwilliger** [comment]
I might be able to take a look at this one, but I have a question in that case.

If we have `Collection<TCollectionDefinition>`, do we also need `CollectionDefinition<TCollectionDefinition>`? Or can the type argument to the generic collection attribute be the reference to the collection fixture?

**Comment by bradwilson** [comment]
@JamesTerwilliger Hmm, good point. I don't think having `[CollectionDefinition<T>]` attached to `T` itself makes much sense. We could just add a parameterless constructor to the existing non-generic attribute and document it to say "use this constructor if you plan to use `[Collection<T>]`" (which itself obviously would not offer a name string, unlike `[Collection]`).

**Comment by bradwilson** [comment]
> I do wonder if we could make it simpler by not even requiring any [CollectionDefinition] at all in this case.

Duh! Of course. :)

**Comment by JamesTerwilliger** [comment]
It does appear that `[CollectionDefinition]` is where one disables parallelization, perhaps it becomes optional in the `[Collection<T>]` case the definition attribute is only required if its properties are used.

**Comment by JamesTerwilliger** [comment]
First attempt: https://github.com/xunit/xunit/pull/2878

Still needs some tests, and need to figure out what the collection key should be in the string-less case. Currently defaulting to ITypeInfo.SimpleName.

**Comment by bradwilson** [comment]
I don't think it's necessary (or even a good idea) to create a "simple" name when using `[Collection<T>]`. In fact I'd prefer the name be unpredictable so that it doesn't conflict with the names people craft by hand (that is, I don't think it's a good idea for people to be able to mix `[Collection("MyClass")]` and `[Collection<MyClass>]` to point to the same collection definition).

You can use [`UniqueIDGenerator`](https://github.com/xunit/xunit/blob/main/src/xunit.v3.common/Utility/UniqueIDGenerator.cs) from `xunit.v3.common` to generate a unique ID for the fixture name. I would probably feed `_ITypeInfo.Assembly.Name` and `_ITypeInfo.Name` into it to get the right level of uniqueness.

**Comment by JamesTerwilliger** [comment]
OK, moved name generation to UniqueIDGenerator, should be ready.

My main missing bit in the PR is that I don't have tests yet for `[Collection<TDefinition>]` because the main test assembly is in 6.0. Where would you like tests to live that need to be in 7.0 or higher?

**Comment by bradwilson** [comment]
> Where would you like tests to live that need to be in 7.0 or higher?

I'll add a `net8.0` target framework for `xunit.v3.core.tests`, and you can gate the tests with `#if NET7_0_OR_GREATER`.

**Comment by bradwilson** [comment]
@JamesTerwilliger Okay, I just pushed up the `net8.0` change. You can merge `main` to get access to it.

**Comment by JamesTerwilliger** [comment]
@bradwilson Perfect- I added some tests for the generic version (which indeed required a change to ReflectionTypeInfo to mimic newer GetCustomAttribute behavior for .Net7.0+) and took the PR out of draft mode.

**Comment by bradwilson** [comment]
Available in `0.1.1-pre.369`
